### PR TITLE
Ignore hasOutstandingIOU from child iouReport

### PR DESCRIPTION
**Pull request by aldo-expensify**
Test with: https://github.com/Expensify/Web-Expensify/pull/38473

### Details

We were looking into the linked iou/expense report to check for the property `hasOutstandingIOU`. This means that if you request money, both reports (the parent chat report and the money request report) get the green dot like there is an action pending.

We are changing this to now only show the green dot in the expense/iou report and not the parent report.

The original behaviour made sense when we were not loading iou/expense reports into the LHN, but we are now, so we don't need this logic anymore

### Fixed Issues

$ https://github.com/Expensify/Expensify/issues/307581
PROPOSAL: 

### Tests

1. Create two accounts
2. Create a chat between the accounts
3. Account1 creates a money request from Account2
4. Verify that Account2 only sees the green dot in the IOU report and not in the chat
5. Log out and log in with Account2
6. Verify again that Account2 only sees the green dot in the IOU report and not in the chat

- [x] Verify that no errors appear in the JS console

### Offline tests

N/A

### QA Steps

1. Create two accounts
2. Create a chat between the accounts
3. Account1 creates a money request from Account2
4. Verify that Account2 only sees the green dot in the IOU report and not in the chat
5. Log out and log in with Account2
6. Verify again that Account2 only sees the green dot in the IOU report and not in the chat

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

**Before**
Screenshot 2023-08-11 at 11 22 33 AM

**After**
Screenshot 2023-08-11 at 11 22 54 AM

</details>

<details>
<summary>Mobile Web - Chrome</summary>

</details>

<details>
<summary>Mobile Web - Safari</summary>

</details>

<details>
<summary>Desktop</summary>

</details>

<details>
<summary>iOS</summary>

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

**Comment by situchan** [comment]
> Test with: https://github.com/Expensify/Web-Expensify/pull/38473

What's the status of this PR? local / merged only / staging?

**Comment by aldo-expensify** [comment]
> What's the status of this PR? local / merged only / staging?

I don't think it is necessary to wait for that, both parts (frontend and backend) needs the same update, and it will be inconsistent as long as one of them has not been deployed

**Comment by aldo-expensify** [comment]
@situchan thinking again about it, I guess it makes sense to wait for https://github.com/Expensify/Web-Expensify/pull/38473 to be deployed, otherwise it will be hard to test that the front end changes are doing something.

File: src/libs/ReportUtils.js [hunk 1]
```diff
@@ -1148,15 +1146,7 @@ function isWaitingForIOUActionFromCurrentUser(report, allReportsDict = null) {
...
-    let reportToLook = report;
-    if (report.iouReportID) {
-        const iouReport = allAvailableReports[`${ONYXKEYS.COLLECTION.REPORT}${report.iouReportID}`];
-        if (iouReport) {
-            reportToLook = iouReport;
-        }
-    }
-    // Money request waiting for current user to Pay (from chat or from iou report)
-    if (reportToLook.ownerAccountID && (reportToLook.ownerAccountID !== currentUserAccountID || currentUserAccountID === reportToLook.managerID) && reportToLook.hasOutstandingIOU) {
+    if (report.hasOutstandingIOU && report.ownerAccountID && (report.ownerAccountID !== currentUserAccountID || currentUserAccountID === report.managerID)) {
```
Comment (mountiny) [inline src/libs/ReportUtils.js:1149]: It would be great to add a comment here to explain what exactly we are checking for here in plain english, or in other words list out the scenarios for when the `isWaitingForIOUActionFromCurrentUser` should return true so contributors can also understand this faster and easier

Comment (aldo-expensify) [reply src/libs/ReportUtils.js:1151]: Are you suggesting to restore the comment: "// Money request waiting for current user to Pay (from chat or from iou report)" ? That makes sense to me

Comment (aldo-expensify) [reply src/libs/ReportUtils.js:1152]: reading you comment again, were you suggesting to add the list of cases this function returns `true` in the documentation of the function?

**Comment by aldo-expensify** [comment]
Fixed tests, the Web-E PR is in staging, so this can be tested now

**Comment by JmillsExpensify** [comment]
@situchan Heads up that this is one of the most important PRs we can merge right now. Please start on this one before any others in your list.

File: src/libs/ReportUtils.js (hunk 1 above)
Comment (mountiny) [reply src/libs/ReportUtils.js:1152]: Yeah haha, basically list out in plain English when the green dot should show

**Comment by mountiny** [comment]
Asked in Slack for prioritization https://expensify.slack.com/archives/C02NK2DQWUX/p1692108871664589

**Comment by situchan** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details open>
<summary>Web</summary>

- IOU request

https://github.com/Expensify/App/assets/108292595/8c423515-5bc2-4c7e-9df7-be1c9c05a70b

- Expense request


https://github.com/Expensify/App/assets/108292595/7194e694-4993-4c00-b54c-60f158e9333d



</details>

<details>
<summary>Mobile Web - Chrome</summary>



</details>

<details>
<summary>Mobile Web - Safari</summary>



</details>

<details>
<summary>Desktop</summary>



</details>

<details>
<summary>iOS</summary>



</details>

<details>
<summary>Android</summary>



</details>

**Comment by situchan** [comment]
@aldo-expensify can you please check https://github.com/Expensify/App/pull/24448#issuecomment-1679108819? Constantly reproducible to me

**Comment by mountiny** [comment]
Hmm should be on staging only, waiting for @aldo-expensify to check this out

**Comment by aldo-expensify** [comment]
Looking into this now!

**Comment by aldo-expensify** [comment]
> While testing expense request, I noticed that green dot briefly appears and then disappears in workspace chat
> 
>  bug.mov

The workspace chat should have the green dot, only the expense report created should have it. I'm not being able to reproduce this.

Regarding the backend PR, it is currently in `staging` only, and it should affect if you see the green dot or not in the workspace chat.

**Comment by situchan** [comment]
> The workspace chat should have the green dot, only the expense report created should have it.

you meant `The workspace chat should not have the green dot`?

**Comment by aldo-expensify** [comment]
> > The workspace chat should have the green dot, only the expense report created should have it.
> 
> you meant `The workspace chat should not have the green dot`?

hehe yes, sorry.. corrected it

**Comment by aldo-expensify** [comment]
I have tested with my dev API, I'll try with staging to see if there is a problem

**Comment by aldo-expensify** [comment]
@situchan maybe I'm wrong about expecting a difference between staging/production, maybe this doesn't depend on the backend update to work fine. Testing...

**Comment by aldo-expensify** [comment]
> > @situchan maybe I'm wrong about expecting a difference between staging/production, maybe this doesn't depend on the backend update to work fine. Testing...
> 
> I also thought so. Though not sure what backend update is.

The backend should make the `policyExpenseChat` have `hasOutstandingIOU: false` since it doesn't copy that state anymore from the linked iou/expense report, but as we saw, seems like that change is not really important for this to work.

**Comment by aldo-expensify** [comment]
> > I'm trying to make my app hit staging using `.env` but it still goes to production 😕
> 
> I am always using this toggle on Preferences page for staging/production switch
> 
> Screenshot 2023-08-16 at 2 04 39 AM

I see... I'm not sure if that works for my case because I think you can activate that after you logged in, and the server was sending the wrong thing in the `OpenApp` request

Anyway, I got it to connect to staging in my env by adding `ENVIRONMENT=staging`, now the `policyExpenseChat` has `hasOutstandingIOU: false`:

image

Not sure if it helps with anything you are testing. Are you able to reproduce the issue with the green dot in the workspace chat? if you can reproduce, can you give me a list of steps?

**Comment by situchan** [comment]
Trying to find the root cause.

This video shows clear reproduction step.


https://github.com/Expensify/App/assets/108292595/46d04757-5b80-4669-b144-03b8545c13b7

**Comment by aldo-expensify** [comment]
> This video shows clear reproduction step.

Not clear enough :P . Does this happen if you do the same with a fresh account?
Is the account requesting money the same as the workspace admin/owner? or is it a member that you invited?

**Comment by situchan** [comment]
I think I found the root cause.
Pusher is still sending `hasOutstandingIOU: true`. As seen in video, when switch to another chat and come back, green dot disappears, which means `hasOutstandingIOU` is set back to false

**Comment by situchan** [comment]
Is it possible that pusher points to production while api points to staging?

I think we can go ahead and merge this as this is related to backend.
And after deploy to staging, re-test

**Comment by aldo-expensify** [comment]
> Is it possible that pusher points to production while api points to staging?

hmm, the pusher event are per account, and they are sent when the command is executed in the backend. If the command reached staging, then the generated pusher events should have been generated with staging code.

Do you think it is a pusher event generated from the `RequestMoney` request? can you give me the `requestID` of the request `RequestMoney` when you are testing with staging? You can get it from here:

image

**Comment by aldo-expensify** [comment]
@situchan I just reproduced the bug you are pointing too, but I had to create a workspace with 2 users and request money from the member that is not the admin. I can see the pusher event that incorrectly sets `hasOutstandingIOU: true` for the `policyExpenseChat` and the green dot appearing:

image

I'll work on fixing that in the backend, thanks for the catch!

**Comment by mountiny** [comment]
Nice, great job identifying it, I dont think we have to hold on that one though so @situchan if you can fill out the checklist even without it that would be great! thanks!

**Review by situchan — APPROVED** [review]
Checklist: https://github.com/Expensify/App/pull/24448#issuecomment-1679080167

**Comment by aldo-expensify** [comment]
> I dont think we have to hold on that one though so I think we can go ahead with this PR, do you agree @aldo-expensify?

Agree, this is a pretty safe change

**Review by mountiny — APPROVED** [review]
Changes look good to me and @situchan tested thoroughly based on the comments. Feel free to merge if you want, I think we are good to go @aldo-expensify
